ipamd/datastore/data_store.go

```diff
@@ -287,22 +304,57 @@ func (ds *DataStore) UnAssignPodIPv4Address(k8sPod *k8sapi.K8SPodInfo) (string,
 	}
 
 	for _, eni := range ds.eniIPPools {
-		ip, ok := eni.ipv4Addresses[ipAddr.IP]
-		if ok && ip.assigned {
-			ip.assigned = false
+		ip, ok := eni.IPv4Addresses[ipAddr.IP]
+		if ok && ip.Assigned {
+			ip.Assigned = false
 			ds.assigned--
-			eni.assignedIPv4Addresses--
+			eni.AssignedIPv4Addresses--
 			curTime := time.Now()
 			ip.unAssignedTime = curTime
-			eni.lastUnassignedTime = curTime
+			eni.lastUnAssignedTime = curTime
 			log.Infof("UnAssignIPv4Address: Pod (Name: %s, NameSpace %s Container %s)'s ipAddr %s, DeviceNumber%d",
-				k8sPod.Name, k8sPod.Namespace, k8sPod.Container, ip.address, eni.deviceNumber)
+				k8sPod.Name, k8sPod.Namespace, k8sPod.Container, ip.address, eni.DeviceNumber)
 			delete(ds.podsIP, podKey)
-			return ip.address, eni.deviceNumber, nil
+			return ip.address, eni.DeviceNumber, nil
 		}
 	}
 
 	log.Warnf("UnassignIPv4Address: Failed to find pod %s namespace %s Container %s using ip %s",
 		k8sPod.Name, k8sPod.Namespace, k8sPod.Container, ipAddr.IP)
 	return "", 0, ErrUnknownPodIP
 }
+
+// GetPodInfos provides pods ip information to introspection endpoint
+func (ds *DataStore) GetPodInfos() *map[string]PodIPInfo {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	var podInfos = make(map[string]PodIPInfo)
+
+	for podKey, podInfo := range ds.podsIP {
+		key := podKey.name + "_" + podKey.namespace + "_" + podKey.container
+		podInfos[key] = podInfo
+		log.Debugf("introspect: key %s", key)
+	}
+
+	log.Debugf("introspect: len %d", len(ds.podsIP))
+
+	return &podInfos
+}
+
+// GetENIInfos provides eni ip information to introspection endpoint
+func (ds *DataStore) GetENIInfos() *ENIInfos {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	var eniInfos = ENIInfos{
+		TotalIPs:    ds.total,
+		AssignedIPs: ds.assigned,
+		ENIIPPools:  make(map[string]ENIIPPool),
```

Comment:
`make(map[string]ENIIPPool, len(ds.eniIPPools))` will give the right initial size and avoid reallocations as you copy.

Reply:
done. thanks